src/cmap/wait_queue.rs

```diff
@@ -1,177 +1,85 @@
-    /// temporarily pass ownership of the guard to `Condvar::wait` (or `Condvar::wait_timeout`),
-    /// which will return the mutex guard when it's done blocking. Since we are unable to take
-    /// direct ownership of a field in a method that does not take ownership, we're forced to make
-    /// the `guard` field an `Option` and temporarily remove the guard from the `Option` while
-    /// waiting.
-    guard: Option<MutexGuard<'a, WaitQueueInner>>,
-
-    /// The conditional variable used to facilitate waiting for a connection to be ready.
-    condvar: Arc<Condvar>,
-}
-
-impl<'a> WaitQueueHandle<'a> {
-    /// Creates a new `WaitQueueHandle`.
-    fn new(guard: MutexGuard<'a, WaitQueueInner>, condvar: Arc<Condvar>) -> Self {
-        Self {
-            guard: Some(guard),
-            condvar,
-        }
-    }
-
-    /// Blocks on the internal conditional variable until either the conditional variable is
-    /// notified (which indicates that a connection is ready) or a timeout occurs.
-    pub(super) fn wait_for_available_connection(
-        &mut self,
-        timeout: Option<Duration>,
-    ) -> Result<()> {
-        // Temporarily remove the mutex guard from the field so that we can pass ownership of it to
-        // the conditional variable.
-        let guard = self.guard.take().unwrap();
-
-        if let Some(timeout) = timeout {
-            // Wait until a connection is checked back in or the timeout has elapsed.
-            let (guard, result) = self.condvar.wait_timeout(guard, timeout).unwrap();
-
-            if result.timed_out() {
-                return Err(ErrorKind::WaitQueueTimeoutError {
-                    address: guard.address.clone(),
-                }
-                .into());
-            }
-
-            self.guard = Some(guard);
-        } else {
-            // Wait until a connection is checked back in.
-            self.guard = Some(self.condvar.wait(guard).unwrap());
-        }
-
-        Ok(())
-    }
-}
 
-impl<'a> Drop for WaitQueueHandle<'a> {
-    /// When the `WaitQueueHandle` goes out of scope, we pop the current thread from the front of
-    /// the wait queue and notify the wait queue that it can wake up the next thread.
-    fn drop(&mut self) {
-        if let Some(ref mut guard) = self.guard {
-            guard.queue.pop_front();
-            guard.notify_ready();
-        }
-    }
+    /// How long a thread is allowed to be waiting in the queue before timing out.
+    timeout: Option<Duration>,
 }
 
 impl WaitQueue {
-    /// Creates a new `WaitQueue`.
-    pub(super) fn new(address: StreamAddress, timeout: Option<Duration>) -> Self {
-        let inner = WaitQueueInner {
-            queue: Default::default(),
-            address,
-            timeout,
-        };
+    /// Creat a new `WaitQueue`.
+    pub(super) fn new(address: StreamAddress, max_handles: u32, timeout: Option<Duration>) -> Self {
+        let max_handles = max_handles.try_into().unwrap_or(usize::max_value());
```

Comment:
If the number of handles doesn't fit into a `usize`, it might make more sense to just assume there's no max pool size; unless the driver is getting run on a 16-bit processor, I doubt that a max pool size of `usize::max_value()` would ever hit its limit.

Reply:
as discussed offline, to ensure fairness we want to keep using the semaphore, so we'll treat `usize::max_value` as essentially no limit. As part of this, I added handling 0 as an indication of no limit to the pool options parsing.